Python/specialize.c

```diff
@@ -2568,28 +2574,40 @@ _Py_Specialize_ContainsOp(PyObject *value, _Py_CODEUNIT *instr)
     assert(ENABLE_SPECIALIZATION);
     assert(_PyOpcode_Caches[CONTAINS_OP] == INLINE_CACHE_ENTRIES_COMPARE_OP);
     _PyContainsOpCache *cache = (_PyContainsOpCache  *)(instr + 1);
-    if (PyUnicode_CheckExact(value)) {
-        instr->op.code = CONTAINS_OP_STR;
-        goto success;
-    }
-    if (PyList_CheckExact(value)) {
-        instr->op.code = CONTAINS_OP_LIST;
-        goto success;
-    }
-    if (PyTuple_CheckExact(value)) {
-        instr->op.code = CONTAINS_OP_TUPLE;
-        goto success;
-    }
     if (PyDict_CheckExact(value)) {
         instr->op.code = CONTAINS_OP_DICT;
         goto success;
     }
-    if (PySet_CheckExact(value)) {
+    if (PySet_CheckExact(value) || PyFrozenSet_CheckExact(value)) {
         instr->op.code = CONTAINS_OP_SET;
         goto success;
     }
 
+    if (Py_TYPE(value) == &PyDictProxy_Type) {
+        instr->op.code = CONTAINS_OP_MAPPINGPROXY;
+        goto success;
+    }
 
+#ifdef Py_STATS
```

Comment:
Can you move this into its own function, like `load_attr_fail_kind`, then this becomes `SPECIALIZATION_FAIL(CONTAINS_OP, contains_fail_kind(value))`